cirq/contrib/quirk/cells/arithmetic_cells.py

```diff
@@ -0,0 +1,316 @@
+                return args[0]
+
+        result = self.func(*args)
+        if self.is_modular:
+            result %= args[-1]
+        return result
+
+
+class ArithmeticCell(Cell):
+
+    def __init__(self, identifier: str, target: Sequence['cirq.Qid'],
+                 inputs: Sequence[Union[None, Sequence['cirq.Qid'], int]]):
+        self.identifier = identifier
+        self.target = target
+        self.inputs = inputs
+
+    @property
+    def operation(self):
+        return ARITHMETIC_OP_TABLE[self.identifier]
+
+    def with_input(self, letter, register):
+        new_inputs = [
+            reg if letter != reg_letter else register
+            for reg, reg_letter in zip(self.inputs, self.operation.letters)
+        ]
+        return ArithmeticCell(self.identifier, self.target, new_inputs)
+
+    def operations(self) -> 'cirq.OP_TREE':
+        missing_inputs = [
+            letter for reg, letter in zip(self.inputs, self.operation.letters)
+            if reg is None
+        ]
+        if missing_inputs:
+            raise ValueError(f'Missing input: {sorted(missing_inputs)}')
+
+        return QuirkArithmeticOperation(
+            self.identifier, self.target,
+            cast(Sequence[Union[Sequence['cirq.Qid'], int]], self.inputs))
+
+
+def _indented_list_lines_repr(items: Sequence[Any]) -> str:
+    block = '\n'.join([repr(op) + ',' for op in items])
+    indented = '        ' + '\n        '.join(block.split('\n'))
+    return '[\n{}\n    ]'.format(indented)
+
+
+def _generate_helper() -> Iterator[CellMaker]:
+    # Comparisons.
+    yield _arithmetic_gate("^A<B", 1, lambda x, a, b: x ^ int(a < b))
+    yield _arithmetic_gate("^A>B", 1, lambda x, a, b: x ^ int(a > b))
+    yield _arithmetic_gate("^A<=B", 1, lambda x, a, b: x ^ int(a <= b))
+    yield _arithmetic_gate("^A>=B", 1, lambda x, a, b: x ^ int(a >= b))
+    yield _arithmetic_gate("^A=B", 1, lambda x, a, b: x ^ int(a == b))
+    yield _arithmetic_gate("^A!=B", 1, lambda x, a, b: x ^ int(a != b))
+
+    # Addition.
+    yield from _arithmetic_family("inc", lambda x: x + 1)
+    yield from _arithmetic_family("dec", lambda x: x - 1)
+    yield from _arithmetic_family("+=A", lambda x, a: x + a)
+    yield from _arithmetic_family("-=A", lambda x, a: x - a)
+
+    # Multiply-accumulate.
+    yield from _arithmetic_family("+=AA", lambda x, a: x + a * a)
+    yield from _arithmetic_family("-=AA", lambda x, a: x - a * a)
+    yield from _arithmetic_family("+=AB", lambda x, a, b: x + a * b)
+    yield from _arithmetic_family("-=AB", lambda x, a, b: x - a * b)
+
+    # Misc.
+    yield from _arithmetic_family("+cntA", lambda x, a: x + _popcnt(a))
+    yield from _arithmetic_family("-cntA", lambda x, a: x - _popcnt(a))
+    yield from _arithmetic_family("^=A", lambda x, a: x ^ a)
+    yield from _arithmetic_family(
+        "Flip<A", lambda x, a: a - x - 1 if x < a else x)
+
+    # Multiplication.
+    yield from _arithmetic_family("*A", lambda x, a: x * a if a & 1 else x)
```

Comment:
This is surprising. Why don't we perform multiplication by a when a is even?

Reply:
Because that's what Quirk does and we have to match the behavior.

Quirk does it because multiplication by 2 is not a reversible operation on a fixed width binary register. The alternative I considered was to replace factors of 2 by a left rotate, but I rejected that because in the modular case it require factoring the modulus and factoring was too expensive to do while remaining performant.